server/src/main/java/io/druid/client/CachingClusteredClient.java

```diff
@@ -265,27 +271,60 @@ public CachingClusteredClient(
       }
 
       final Set<ServerToSegment> segments = computeSegmentsToQuery(timeline);
-      @Nullable final byte[] queryCacheKey = computeQueryCacheKey();
+      @Nullable
+      final byte[] queryCacheKey = computeQueryCacheKey();
+      @Nullable
+      final String queryResultKey = computeCurrentEtag(segments, queryCacheKey);
       if (query.getContext().get(QueryResource.HEADER_IF_NONE_MATCH) != null) {
-        @Nullable final String prevEtag = (String) query.getContext().get(QueryResource.HEADER_IF_NONE_MATCH);
-        @Nullable final String currentEtag = computeCurrentEtag(segments, queryCacheKey);
+        @Nullable
+        final String prevEtag = (String) query.getContext().get(QueryResource.HEADER_IF_NONE_MATCH);
+        @Nullable
+        final String currentEtag = queryResultKey;
         if (currentEtag != null && currentEtag.equals(prevEtag)) {
           return Sequences.empty();
         }
       }
 
+      @Nullable
+      final byte[] cachedResultSet = fetchFromResultLevelCache(queryResultKey);
+      if (cachedResultSet != null) {
+        log.info("Fetching entire result set from cache");
+        return fetchSequenceFromResultLevelCache(cachedResultSet);
+      }
       final List<Pair<Interval, byte[]>> alreadyCachedResults = pruneSegmentsWithCachedResults(queryCacheKey, segments);
       final SortedMap<DruidServer, List<SegmentDescriptor>> segmentsByServer = groupSegmentsByServer(segments);
-      return new LazySequence<>(() -> {
+      return Sequences.wrap(new LazySequence<>(() -> {
```

Comment:
How does this impact the lazy evaluation? Where is the cache work done? in the jetty thread?

Reply:
Lazy evaluation would continue to work as it is. The wrapped after() method would run after all the results have been accumulated and populates the cache in the jetty thread.